arangod/Aql/OptimizerRules.cpp

```diff
@@ -6125,36 +6141,105 @@ void arangodb::aql::optimizeTraversalsRule(Optimizer* opt,
     // later (note that the path out variable can contain the "vertices" sub
     // attribute)
     auto outVariable = traversal->vertexOutVariable();
+    if (outVariable != nullptr) {
+      if (!n->isVarUsedLater(outVariable) &&
+          std::find(pruneVars.begin(), pruneVars.end(), outVariable) ==
+              pruneVars.end()) {
+        outVariable = traversal->pathOutVariable();
+        if (outVariable == nullptr ||
+            ((!n->isVarUsedLater(outVariable) ||
+              !options->producePathsVertices()) &&
+             std::find(pruneVars.begin(), pruneVars.end(), outVariable) ==
+                 pruneVars.end())) {
+          // both traversal vertex and path outVariables not used later
+          options->setProduceVertices(false);
+          modified = true;
+        }
+      }
+    }
 
-    if (outVariable != nullptr && !n->isVarUsedLater(outVariable) &&
-        std::find(pruneVars.begin(), pruneVars.end(), outVariable) ==
+    outVariable = traversal->edgeOutVariable();
+    if (outVariable != nullptr) {
+      if (!n->isVarUsedLater(outVariable)) {
+        // traversal edge outVariable not used later
+        options->setProduceEdges(false);
+        if (std::find(pruneVars.begin(), pruneVars.end(), outVariable) ==
             pruneVars.end()) {
-      outVariable = traversal->pathOutVariable();
-      if (outVariable == nullptr ||
-          (!n->isVarUsedLater(outVariable) &&
-           std::find(pruneVars.begin(), pruneVars.end(), outVariable) ==
-               pruneVars.end())) {
-        // both traversal vertex and path outVariables not used later
-        options->setProduceVertices(false);
+          traversal->setEdgeOutput(nullptr);
+        }
         modified = true;
       }
     }
 
-    outVariable = traversal->edgeOutVariable();
-    if (outVariable != nullptr && !n->isVarUsedLater(outVariable)) {
-      // traversal edge outVariable not used later
-      options->setProduceEdges(false);
-      if (std::find(pruneVars.begin(), pruneVars.end(), outVariable) ==
-          pruneVars.end()) {
-        traversal->setEdgeOutput(nullptr);
+    // handle projections (must be done after path variable optimization)
```

Comment:
I think it could, would like to avoid it for now, to get this through.
Adding new Optimizer Rules is a bit more work to do.

Reply:
It could just be a function that is called from here to make the code look sligthly less spaghetti-ish; making it into an optimiser rule is a step which is out of scope for this PR.